### PR TITLE
Feat: add PDF input support with PyMuPDF

### PR DESCRIPTION
**Pull request by violence9331**
## Summary
- Scan the. pdf file as OMR input, convert the PDF page into a grayscale numpy array using PyMuPDF (fitz), and reuse the existing detection pipeline
- Add the config. json configuration item pdf_params to control rendering DPI and page selection
- Save the middle image in. png format, overwritting will raise warning
- Multi page PDF is processed independently page by page, producing independent results for each page

## Main changes
| File Path | Description |
| :--- | :--- |
| `src/entry.py` | **Core Changes:** Added `load_omr_image()` function to render PDFs into `numpy` arrays using `fitz` (PyMuPDF); added `*.pdf` to supported extensions (`exts`); adapted `process_files` and `show_template_layouts` to support multi-page PDFs. |
| `samples/` | Added sample files in PDF format for sample1&2. Sample 2 is for testing multi page processing |

## About Multi page PDF
Now each page is processed independently, treating all pages as the same test paper. I noticed the idea in the previous issue #145, perhaps putting different pages of the test paper in PDF is a solution to image format matching, but I did not make further modifications here. If this implementation method is not necessary, please notify me

### CODE REVIEW

File: src/entry.py [hunk 1]
```diff
@@ -185,156 +186,215 @@ def process_dir(
...
 
-        outputs_namespace.OUTPUT_SET.append([file_name] + resp_array)
+    score = 0
+    if evaluation_config is not None:
+        score = evaluate_concatenated_response(
+            omr_response,
+            evaluation_config,
+            file_path,
+            outputs_namespace.paths.evaluation_dir,
+        )
```
[the remark this answers is not in the document]
Comment (Udayraj123) [reply src/entry.py:298]: @violence9331 what do you think about this? can we add validation/notes in the CLI help or readme about the usage for pdfs?

File: samples/sample1/MobileCamera/sheet1.pdf
Comment (Udayraj123) [inline samples/sample1/MobileCamera/sheet1.pdf:1]: can we keep the pdf size under 300KB?

File: src/defaults/config.py [hunk 2]
```diff
@@ -24,6 +24,10 @@
             "stride": 1,
             "thickness": 3,
         },
+        "pdf_params": {
+            "pdf_dpi": 200,
+            "pdf_page": None,
```
Comment (Udayraj123) [inline src/defaults/config.py]: why not keep default page as page 1?

File: src/entry.py [hunk 3]
```diff
@@ -185,156 +186,232 @@ def process_dir(
         )
 
 
+def load_omr_image(file_path, tuning_config):
+    """Load OMR image from file. Returns list of (display_name, image_array) tuples."""
+    suffix = file_path.suffix.lower()
+    if suffix == ".pdf":
```
Comment (Udayraj123) [inline src/entry.py]: let's move this pdf logic into a util function

File: src/entry.py [hunk 4]
```diff
@@ -185,156 +186,232 @@ def process_dir(
...
+            pdf_params = tuning_config.pdf_params
+            page_num = pdf_params.pdf_page
+            if page_num is not None and page_num >= len(doc):
+                logger.warning(
+                    f"PDF page {page_num} out of range for '{file_path}' "
+                    f"(has {len(doc)} pages), skipping."
+                )
+                return []
+
+            pages = [page_num] if page_num is not None else range(len(doc))
```
Comment (Udayraj123) [inline src/entry.py]: how about supporting an array of page numbers as part of the schema as well?

**Review by Udayraj123 — COMMENTED** [review]
thanks for picking up a long-due issue. Please check the comments and also add a recording/screenshot of the sample tests working.

File: samples/sample1/MobileCamera/sheet1.pdf
Comment (violence9331) [reply samples/sample1/MobileCamera/sheet1.pdf:1]: no problem

File: src/defaults/config.py (hunk 2 above)
Comment (violence9331) [reply src/defaults/config.py]: It will be updated to "pdf_dpi": "automatic", "pdf_page": 1 and will be introduced later

File: src/entry.py (hunk 3 above)
Comment (violence9331) [reply src/entry.py]: done

**Comment by violence9331** [comment]
## Updates & Optimizations
- Regarding file size: Previously, lossless embedding of PDFs would cause the file size to exceed the image size, but now lossy embedding is used to maintain file size
- After updating the file, it was found that interpolating the image according to the default dpi rendering may cause the image edges to soften and not recognize rectangles. The current default "auto" dpi selection will automatically match the dpi to avoid distortion
- "pdf_page" is also compatible with string and array: when using it, you can use "1-5", [1,3,5] or a combination like [1,"3-5"], and it will perform out of bounds detection and warning
## Test cases
All tests are converted from the original images, I will verify the results they have identified and take screenshots of some of them
<details>
<summary>Sample1</summary>
This is a single page PDF, also test for same name situation and overwrite warning, I changed the extension of the jpg file to png, and it was changed back after testing
image
This is a normal result:

| file_id | input_path | output_path | score | Roll | q1 | q2 | q3 | q4 | q5 | q6 | q7 | q8 | q9 | q10 | q11 | q12 | q13 | q14 | q15 | q16 | q17 | q18 | q19 | q20 |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| sheet1.jpg | samples\sample1\MobileCamera\sheet1.jpg | outputs\test\sample1\MobileCamera\CheckedOMRs\sheet1.jpg | 0 | E503110026 | B | | D | B | 6 | 11 | 20 | 7 | 16 | B | D | C | D | A | D | B | A | C | C | D |
| sheet1.png | samples\sample1\MobileCamera\sheet1.pdf | outputs\test\sample1\MobileCamera\CheckedOMRs\sheet1.png | 0 | E503110026 | B | | D | B | 6 | 11 | 20 | 7 | 16 | B | D | C | D | A | D | B | A | C | C | D |

</details>

<details>
<summary>Sample2</summary>

Sample2 test parameter "pdf_page": "1-2"

| file_id | input_path | output_path | score | q1 | q2 | q3 | q4 | q5 |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| adrian_omr.png | samples\sample2\AdrianSample\adrian_omr.png | ...\CheckedOMRs\adrian_omr.png | 0 | B | E | A | C | B |
| adrian_omr_2.png | samples\sample2\AdrianSample\adrian_omr_2.png | ...\CheckedOMRs\adrian_omr_2.png | 0 | C | E | A | B | B |
| **AdrianSample_p1.png** | samples\sample2\AdrianSample.pdf | ...\CheckedOMRs\AdrianSample_p1.png | 0 | B | E | A | C | B |
| **AdrianSample_p2.png** | samples\sample2\AdrianSample.pdf | ...\CheckedOMRs\AdrianSample_p2.png | 0 | C | E | A | B | B |

Also test for range out "1-3":
image


</details>

<details>
<summary>Sample3</summary>

Sample 3 test parameter "pdf_page": [1,3]:

| file_id | input_path | output_path | score | q1 | q2 | q3 | q4 | q5 | q6 | q7 | q8 | q9 | q10 | q11 |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **AdrianSample_p1.png** | samples\sample4\AdrianSample.pdf | ...\AdrianSample_p1.png | **33.0** | B | D | C | B | D | C | BC | A | C | D | C |
| **AdrianSample_p3.png** | samples\sample4\AdrianSample.pdf | ...\AdrianSample_p3.png | **-2.0** | A | | D | C | AC | A | D | B | C | D | D |
| IMG_20201116_143512.jpg | samples\sample4\IMG_20201116_143512.jpg | ...\IMG_20201116_143512.jpg | 33.0 | B | D | C | B | D | C | BC | A | C | D | C |
| IMG_20201116_150717658.jpg | samples\sample4\IMG_20201116_150717658.jpg | ...\IMG_20201116_150717658.jpg | 33.0 | B | D | C | B | D | C | BC | A | C | D | C |
| IMG_20201116_150750830.jpg | samples\sample4\IMG_20201116_150750830.jpg | ...\IMG_20201116_150750830.jpg | -2.0 | A | | D | C | AC | A | D | B | C | D | D |

</details>

**Comment by violence9331** [comment]
Hi @Udayraj123, friendly ping on this PR when you have a moment. I've addressed your previous feedback and would love to hear your thoughts whenever it's convenient. Thanks!

File: samples/sample4/multiple-images.pdf
Comment (Udayraj123) [inline samples/sample4/multiple-images.pdf:1]: this one is a large file as well, possible to compress without affecting output?

File: samples/sample1/MobileCamera/sheet1.pdf
Comment (Udayraj123) [inline samples/sample1/MobileCamera/sheet1.pdf:1]: let's not bloat sample1 with this feature demo. I think a pdf in sample2 is enough, can you remove this?

File: src/entry.py [hunk 5]
```diff
@@ -185,156 +185,184 @@ def process_dir(
...
-        InteractionUtils.show(
-            f"Template Layout: {file_name}", template_layout, 1, 1, config=tuning_config
-        )
+        images = ImageUtils.load_omr_image(file_path, tuning_config)
+        for img_name, in_omr in images:
+            in_omr = template.image_instance_ops.apply_preprocessors(
+                str(file_path), in_omr, template
+            )
+            if in_omr is None:
+                continue
```
Comment (Udayraj123) [inline src/entry.py:215]: why continue? we should mark this file's images into error bucket?

File: src/entry.py [hunk 6]
```diff
@@ -185,156 +185,184 @@ def process_dir(
...
+                in_omr, template, shifted=False, border=2
+            )
+            InteractionUtils.show(
+                f"Template Layout: {img_name}", template_layout, 1, 1, config=tuning_config
+            )
 
 
-def process_files(
-    omr_files,
+def _process_single_image(
```
Comment (Udayraj123) [inline src/entry.py:224]: thanks for trying to breakdown this large function. Can you please add/update tests for the changes done in this?  especially the error/edge cases?

**Review by Udayraj123 — COMMENTED** [review]
Thanks for the contribution. Although I have not made a full spec on what to support, this is a great start

File: samples/sample4/multiple-images.pdf
Comment (violence9331) [reply samples/sample4/multiple-images.pdf:1]: now it comes to 480kb and the result is same

File: samples/sample1/MobileCamera/sheet1.pdf
Comment (violence9331) [reply samples/sample1/MobileCamera/sheet1.pdf:1]: ok

File: src/entry.py (hunk 5 above)
Comment (violence9331) [reply src/entry.py:215]: I followed the error handling flow from other function — files that fail are marked in the error bucket (logged to ErrorFiles.csv + copied to ErrorFiles/ dir). 
Also try to implement the check_and_move function, which was previously a TODO stub.

File: src/entry.py (hunk 6 above)
Comment (violence9331) [reply src/entry.py:224]: I am not sure which you prefer: add some tests in `/tests/test_edge_cases.py` (for integration/unit test) or process a sample with bad images, snapshot the output. 
Happy to adjust to whatever pattern you prefer

Comment (Udayraj123) [reply src/entry.py:224]: I'd like to keep a test around a combination of bad + good image inside the same pdf(say single_image_error.pdf) which gets tested as part of a sample.

File: src/entry.py (hunk 5 above)
Comment (Udayraj123) [reply src/entry.py:215]: Let's not change `check_and_move` as part of this PR, we can take care of it separately with a better spec.

File: src/entry.py (hunk 6 above)
Comment (violence9331) [reply src/entry.py:224]: Hi, I finished testing PDF error handling. Here's what I found:

I put together a PDF with two pages: page 1 is the original sheet1.jpg, page 2 is the same image with GaussianBlur applied. Following is the result:
**Page 1 (Good image)** — Processed successfully, correct recognition:
| file_id | score | Roll | q1-q20 |
|---------|-------|------|--------|
| single_image_error_p1.png | 0 | E503110026 | B,,D,B,6,11,20,7,16,B,D,C,D,A,D,B,A,C,C,D |

**Page 2 (Blurred image)** — CropPage could not find paper boundary, correctly recorded in ErrorFiles:

```csv
file_id,input_path,output_path
single_image_error_p2.png,samples\sample1\single_image_error.pdf,outputs\Manual\ErrorFiles\single_image_error_p2.png
```
There is the screenshot of error message:
image
**Fix Applied**
Added a None check in src/core.py:apply_preprocessors() to break out of the preprocessor loop when any preprocessor returns None, preventing downstream AttributeError crashes.

I also withdrew the previous check_and_move function modification and add some instruction for pdf in readme

Comment (Udayraj123) [reply src/entry.py:224]: I'll be able to spend time on this review after 2-3 weeks. If I don't update, can you please followup?

Comment (violence9331) [reply src/entry.py:224]: Thanks, I will ping you in a few weeks if needed.

File: samples/sample4/multiple-images.pdf
Comment (Udayraj123) [inline samples/sample4/multiple-images.pdf:1]: let's rename this multiple-images.pdf

File: src/entry.py (hunk 5 above)
Comment (Udayraj123) [reply src/entry.py:215]: @violence9331 can you document what happens now when there's an error in an image?

**Review by Udayraj123 — APPROVED** [review]
LGTM, please update your branch with latest master and then document 2 things
- The error case behaviour 
- The output/screenshot of all tests passing (apologies for needing to do it manually as the Github CI is broken as of now - I am refactoring certain things)

**Comment by violence9331** [comment]
Rename is finished. And to make the test compatible with my platform, there are two small test fixes:

1. test_edge_cases.py — The CSV filename `Results_05AM.csv` is timezone-dependent (uses `strftime("%I%p", localtime())`). It was hardcoded to UTC+5:30 . Changed to compute it dynamically using the same freeze_time context the pipeline uses.

2. test_template_validations.py — The error message assertion had a hardcoded forward-slash path (`src/tests/test_samples/sample1`), which breaks on Windows (`\`). Changed to use `str(BASE_SAMPLE_PATH)` so the platform handles the separator.

Here is a screenshot of the test passed:
image


About the error image, CropPage / CropOnMarkers return None when markers aren't found. 
Before: the loop in core.py:42-43 kept going — next preprocessor got None, 
called .shape on it, AttributeError crash.

After (core.py:44-45): break on None, return it to entry.py:247 which already 
handles the None → ErrorFiles.csv path at entry.py:247-254.

Using sample1 to explain, before:
image
After:
image

**Comment by Udayraj123** [comment]
Thanks for making the efforts to deliver this feature! If you have a short message to post in Discord #dev-users channel to feel free to post and get early feedback.

**Comment by Udayraj123** [comment]
Will include this in dev branch(v2) separately as part of my refactoring efforts(probably in a few months)
